### PR TITLE
KEYCLOAK-17446 Prefill username in "Forgot Your Password" form from Login form

### PR DESCRIPTION
**Pull request by velias**
Really small PR to improve UX a bit - username in "Forgot Your Password" form is prefilled by username attempted by user in Login form.

### CODE REVIEW

**Comment by velias** [comment]
@pedroigor can you look at this please? thx

File: themes/src/main/resources/theme/base/login/login-reset-password.ftl [hunk 1]
```diff
@@ -9,8 +9,8 @@
                     <label for="username" class="${properties.kcLabelClass!}"><#if !realm.loginWithEmailAllowed>${msg("username")}<#elseif !realm.registrationEmailAsUsername>${msg("usernameOrEmail")}<#else>${msg("email")}</#if></label>
                 </div>
                 <div class="${properties.kcInputWrapperClass!}">
-                    <#if auth?has_content && auth.showUsername()>
-                        <input type="text" id="username" name="username" class="${properties.kcInputClass!}" autofocus value="${auth.attemptedUsername}" aria-invalid="<#if messagesPerField.existsError('username')>true</#if>"/>
+                    <#if auth?has_content>
```
Comment (pedroigor) [inline themes/src/main/resources/theme/base/login/login-reset-password.ftl:12]: Looks like the `has_content` check is now unnecessary. The bean should be always there.

File: services/src/main/java/org/keycloak/forms/login/freemarker/model/AuthenticationContextBean.java [hunk 2]
```diff
@@ -63,7 +63,7 @@ public String getAttemptedUsername() {
         String username = context.getAuthenticationSession().getAuthNote(AbstractUsernameFormAuthenticator.ATTEMPTED_USERNAME);
 
         // Fallback to real username of the user just if attemptedUsername doesn't exist
-        if (username == null) {
+        if (username == null && context.getUser() != null) {
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/forms/login/freemarker/model/AuthenticationContextBean.java:66]: Only for defensive purposes or did you reach any NPE?

**Review by pedroigor — COMMENTED** [review]
@mposolda Do you remember why the `auth.showUsername()` was added by https://issues.redhat.com/browse/KEYCLOAK-12426.

The proposal here makes sense to me. But I'm not sure why this logic was added.

**Comment by mposolda** [comment]
> @mposolda Do you remember why the `auth.showUsername()` was added by https://issues.redhat.com/browse/KEYCLOAK-12426.
> 
> The proposal here makes sense to me. But I'm not sure why this logic was added.

@pedroigor If you mean why the username is added to the login forms, the motivation is to have it available is described in https://issues.redhat.com/browse/KEYCLOAK-12426 . The idea was to have something like Google is doing - especially for the cases when the first screen of the login form is "UsernameForm" and then password (and other forms) are a follow-up.

The `auth.showUser` is used from the template.ftl - the base template for all login forms. Hence it is used in cases when user is already known (EG. PasswordForm, OTPForm) and also from the cases when user is not known (EG. UsernamePasswordForm) and that is why this check is needed.

**Comment by pedroigor** [comment]
> @pedroigor If you mean why the username is added to the login forms, the motivation is to have it available is described in https://issues.redhat.com/browse/KEYCLOAK-12426 . The idea was to have something like Google is doing - especially for the cases when the first screen of the login form is "UsernameForm" and then password (and other forms) are a follow-up.

Sorry, I was looking for the reason why it was added to `login-reset-password.ftl`. This PR is removing `showUsername` and basically reverting the change from KEYCLOAK-12426.

File: services/src/main/java/org/keycloak/forms/login/freemarker/model/AuthenticationContextBean.java (hunk 2 above)
Comment (velias) [reply services/src/main/java/org/keycloak/forms/login/freemarker/model/AuthenticationContextBean.java:66]: I really reached NPE so I added this check

File: themes/src/main/resources/theme/base/login/login-reset-password.ftl (hunk 1 above)
Comment (velias) [reply themes/src/main/resources/theme/base/login/login-reset-password.ftl:12]: I don't know, I tried to do as small change as necessary.

**Comment by velias** [comment]
> Sorry, I was looking for the reason why it was added to `login-reset-password.ftl`. This PR is removing `showUsername` and basically reverting the change from KEYCLOAK-12426.

To be honest, this PR is not reverting KEYCLOAK-12426, it extends that change. KEYCLOAK-12426 added ability to automatically prefill username in reset password form, but restricted it only to cases when `auth.showUsername()` is `true` (maybe @mposolda remembers why he restricted it this way?). And this PR extends cases when the username is prefilled in this form to show there even attempted username from the login form, which for me is good UX improvement as users reset passwords after unsuccessful login attempts and I do not see reason to force them to enter it manually again. 
We have this kind of UX improvement in sso.redhat.com also.

**Comment by pedroigor** [comment]
@velias It is partially reverting the changes as the check you are removing did not exist. I'm OK with these changes, but I would like to confirm why it was added in the first place. For me, it does not make sense.

File: themes/src/main/resources/theme/base/login/login-reset-password.ftl (hunk 1 above)
Comment (pedroigor) [reply themes/src/main/resources/theme/base/login/login-reset-password.ftl:12]: Sure, but the `else` is unnecessary as per your changes.

**Comment by mposolda** [comment]
@velias @pedroigor Regarding the check of `auth.showUsername` in the `log-reset-password.ftl`, I can't remember any particular reason from the top of my head... Not 100% sure if it is mistake (maybe copy/paste leftover or something like this...). If all the tests are passing, then it is probably ok to remove it as you did in the PR :-)

So the changes in `login-reset-password.ftl` looks good to me. With the only small exception, that if you are confident, that `<#if auth?has_content>` is always true, then it makes sense to remove if/else entirely and simplify this construct. Leaving this to you guys to decide on this...

**Comment by velias** [comment]
Looks like `auth` bean is always present as it is put into context in `org.keycloak.forms.login.freemarker.FreeMarkerLoginFormsProvider.createCommonAttributes()`. So I removed that `if` completely and we will see if all tests pass ;-)

**Review by pedroigor — APPROVED** [review]

**Review by pedroigor — APPROVED** [review]

**Pull request merged**